plumbing/object/commit_walker.go

```diff
@@ -181,3 +183,127 @@ func (w *commitPostIterator) ForEach(cb func(*Commit) error) error {
 }
 
 func (w *commitPostIterator) Close() {}
+
+// commitAllIterator stands for commit iterator for all refs.
+type commitAllIterator struct {
+	// el points to the current commit.
+	el *list.Element
+}
+
+// NewCommitAllIter returns a new commit iterator for all refs.
+// s is a repo Storer used to get commits and references.
+// fn is a commit iterator function, used to iterate through ref commits in chosen order
+func NewCommitAllIter(s storage.Storer, fn func(*Commit) CommitIter) (CommitIter, error) {
+	l := list.New()
+	m := make(map[plumbing.Hash]*list.Element)
+
+	// ...along with the HEAD
+	head, err := storer.ResolveReference(s, plumbing.HEAD)
+	if err != nil {
+		return nil, err
+	}
+	headCommit, err := GetCommit(s, head.Hash())
+	if err != nil {
+		return nil, err
+	}
+	err = fn(headCommit).ForEach(func(c *Commit) error {
+		el := l.PushBack(c)
+		m[c.Hash] = el
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	refIter, err := s.IterReferences()
+	if err != nil {
+		return nil, err
+	}
+	defer refIter.Close()
+	err = refIter.ForEach(func(r *plumbing.Reference) error {
+		if r.Hash() == head.Hash() {
+			// we already have the HEAD
+			return nil
+		}
+		c, _ := GetCommit(s, r.Hash())
+		// if it's not a commit - skip it.
+		if c == nil {
+			return nil
+		}
+
+		el, ok := m[c.Hash]
+		if ok {
+			return nil
+		}
+
+		var refCommits []*Commit
+		cit := fn(c)
+		for c, e := cit.Next(); e == nil; {
+			el, ok = m[c.Hash]
+			if ok {
+				break
+			}
+			refCommits = append(refCommits, c)
```

Comment:
Is this storing all commits in memory? In big repositories this can be a lot.

Reply:
kind of ... it's storing ref commits till it finds an existing hash in the list, then it inserts all memorised commits into the list and clears the slice, but generally speaking the list `l` is a flatten list of all commits ordered by `commitIterFunc`, so yes it can be a lot. It's a first approach, but maybe we should think how to make it in more optimal way. Actually all our iterators don't retrieve any data on demand but just return an object with some offset already stored in memory, e.g.: `Next() { return stack[offset] }`

Reply:
What mean is that you probably only need to store the hash, not the whole `Commit` data. I believe it's not used at all:

```go
type Commit struct {
    Hash Hash
    Author Signature
    Committer Signature
    PGPSignature string
    Message string
    TreeHash Hash
    ParentHashes []Hash
    s EncodedObjectStorer
}
```

Reply:
I think it's not really just Hash, because at the end we need to return `CommitIter` which satisfies:
```
func Next() (*Commit, error)
```

What I can see as a semi-optimization is to keep in memory just hashes, but in `Next()` just call `GetCommit`.

Reply:
Or move this code (the whole iteration) to `Next` so `GetCommit` is only called once.

Reply:
But we need to have flatten list firs, before we start returning `Next` commit. At the beginning you don't know which ref inserts own commits where (also it depends on order)

Reply:
After an offline discussion it seems there's not a straightforward way to create this list on the fly. The biggest problem is the order. Also storing all commit structs in memory may not occupy as much as I previously thought.